train/bleu.cs

```diff
@@ -0,0 +1,199 @@
+                }
+            }
+        }
+
+        public Dictionary<string, int> NgramToDict(int n)
+        //save the ngram to dictionary
+        {
+            Dictionary<string, int> ngram = new Dictionary<string, int>();
+            if (n == 1)
+            {
+                var SentencePiece = OriginalSentence.Split(" ");
+                for (int i = 0; i < SentencePiece.Length; i++)
+                {
+                    if (!ngram.ContainsKey(SentencePiece[i]))
+                    {
+                        ngram[SentencePiece[i]] = 1;
+                    }
+                    else
+                    {
+                        ngram[SentencePiece[i]] += 1;
+                    }
+                }
+                return ngram;
+            }
+            else
+            {
+                IEnumerable<string> result = ngram_queue(n);
+                foreach (string temp in result)
+                {
+                    if (!ngram.ContainsKey(temp))
+                    {
+                        ngram[temp] = 1;
+                    }
+                    else
+                    {
+                        ngram[temp] += 1;
+                    }
+                }
+                return ngram;
+            }
+        }
+    }
+
+    public static void bleu(Ngram N1, Ngram N2)
+    {
+        List<Dictionary<string, int>> N1gram = new List<Dictionary<string, int>>();
+        List<Dictionary<string, int>> N2gram = new List<Dictionary<string, int>>();
+        for (int i = 1; i <= 4; i++)
+        {
+            N1gram.Add(N1.NgramToDict(i));
+            N2gram.Add(N2.NgramToDict(i));
+        }
+        Dictionary<string, double> result = compareTwoGram(N1gram, N2gram);
+        foreach (string i in result.Keys)
+        {
+            Console.WriteLine("p_{0} is {1}", i, result[i]);
+        }
+    }
+    public static Dictionary<string, double> compareTwoGram(List<Dictionary<string, int>> Dict_list1, List<Dictionary<string, int>> Dict_list2)
+    {
+        Dictionary<string, double> result = new Dictionary<string, double>();
+        double BP = 0.00;
+        double logPn = 0.00;
+        //penalty
+        for (int m = 0; m <= 3; m++)
+        {
+            //1,2,3,4gram
+            int count_clip = 0;
+            //save the number of correct prediction
+            int count = 0;
+            //save the number of all grams
+            int temp_length = 0;
+            double temp = 0.0000;
+            //save the P_i possibility
+            Dictionary<string, int> Dict1 = Dict_list1[m];
+            Dictionary<string, int> Dict2 = Dict_list2[m];
+            foreach (string key in Dict1.Keys)
+            {
+                //Console.WriteLine(key);
+                if (Dict2.ContainsKey(key))
+                {
+                    count_clip += Math.Min(Dict1[key], Dict2[key]);
+                    //modified precision
+                }
+            }
+            foreach (int i in Dict1.Values)
```

Comment:
I think you can count the number of grams in your foreach part from line-140 to line-147, so that you could reduce the iteration times. e.g.
(from line 140)
foreach (string key in Dict1.Keys)
{
    count += Dict1[key];
    if (Dict2.ContainsKey(key))
    {
    ......
    }
    ......
}